### PR TITLE
Publicize search by PackageType

### PR DESCRIPTION
**Pull request by ryuyu**
Updates the docs with information about packageType enhanced search and results.
Also updates catalog doc to reflect packageType information now available in catalog.

### CODE REVIEW

**Comment by anangaur** [comment]
Docs Build status updates of commit _[e30d658](https://github.com/NuGet/docs.microsoft.com-nuget/commits/e30d6589da4ff0ef252ff6eccd31179bbd577a5f)_: 

:clock10: **Pending**: waiting for processors ([49](https://ops.microsoft.com/#/statistic) builds ahead of you) 

 :warning: **Docs Build is busy**, currently there are 49 builds ahead of this one, for more information you can view the Build queue graph on the [Docs Portal](https://ops.microsoft.com/#/statistic).

**Comment by anangaur** [comment]
Docs Build status updates of commit _[e30d658](https://github.com/NuGet/docs.microsoft.com-nuget/commits/e30d6589da4ff0ef252ff6eccd31179bbd577a5f)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/api/_data/search-result.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/_data/search-result.json) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-query-service-resource?branch=pr-en-us-2076) |
[docs/api/catalog-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/catalog-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/catalog-resource?branch=pr-en-us-2076) |
[docs/api/search-autocomplete-service-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/search-autocomplete-service-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-autocomplete-service-resource?branch=pr-en-us-2076) |
[docs/api/search-query-service-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/search-query-service-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-query-service-resource?branch=pr-en-us-2076) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2020%5C6%5C23%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C202006230131560106-2076%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=XMs0w4IfG5sMdzX9oBJO2kKWd80b1kLubWpyFdLUzHQ%3D&st=2020-06-23T01%3A27%3A34Z&se=2020-07-24T01%3A32%3A34Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/19a7edc0-f91d-0fd5-6e5f-4d131db05c11?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

File: docs/api/catalog-resource.md [hunk 1]
```diff
@@ -244,6 +245,15 @@ time before the catalog item's commit timestamp.
 The `packageHashAlgorithm` is a string defined by the server implementation representing the hashing algorithm used to
 produce the `packageHash`. nuget.org always used the `packageHashAlgorithm` value of `SHA512`.
 
+The `packageTypes` property will only be present if a package type was specified by the author. If it is present, it will always have at least one (1) entry. Each item in the `packageTypes` array is a JSON object with the following properties:
+
+Name      | Type    | Required | Notes
+--------- | ------- | -------- | -----
+@id       | string  | yes      | The URL to fetch the package type item
+@type     | string  | yes      | Type of package type. Currently will always be "PackageType"
```
Comment (joelverhagen) [inline docs/api/catalog-resource.md:253]: Let's leave `@id` and `@type` undocumented. They are useless and ideally we could drop them with no semantic impact.

File: docs/api/catalog-resource.md [hunk 2]
```diff
@@ -244,6 +245,15 @@ time before the catalog item's commit timestamp.
...
+name      | string  | yes      | The name of the package type.
+verion    | string  | no       | The version of the package type. Only present if the author explicitly specified a version in the nuspec.
```
Comment (joelverhagen) [inline docs/api/catalog-resource.md:255]: spelling: `version`

File: docs/api/search-autocomplete-service-resource.md [hunk 3]
```diff
@@ -69,6 +70,10 @@ If `semVerLevel=2.0.0` is provided, both SemVer 1.0.0 and SemVer 2.0.0 compatibl
 [SemVer 2.0.0 support for nuget.org](https://github.com/NuGet/Home/wiki/SemVer2-support-for-nuget.org-%28server-side%29)
 for more information.
 
+The `packageType` parameter is used to futher filter the search results to only results that are of the provided type.
+If the provided package type is not a valid package type as defined by the [Package Type document](https://github.com/NuGet/Home/wiki/Package-Type-%5BPacking%5D), an empty result will returned.
+If the provided package type is empty, no filter will be applied. In other words, passing no value to the packageType parameter will behave as if the parameter was not passed.
```
Comment (joelverhagen) [inline docs/api/search-autocomplete-service-resource.md:75]:
```suggestion
If the provided package type is empty, no filter will be applied. In other words, passing no value to the `packageType` parameter will behave as if the parameter was not passed.
```

File: docs/api/search-autocomplete-service-resource.md [hunk 4]
```diff
@@ -69,6 +70,10 @@ If `semVerLevel=2.0.0` is provided, both SemVer 1.0.0 and SemVer 2.0.0 compatibl
 [SemVer 2.0.0 support for nuget.org](https://github.com/NuGet/Home/wiki/SemVer2-support-for-nuget.org-%28server-side%29)
 for more information.
 
+The `packageType` parameter is used to futher filter the search results to only results that are of the provided type.
```
Comment (joelverhagen) [inline docs/api/search-autocomplete-service-resource.md:73]:
```suggestion
The `packageType` parameter is used to futher filter the search results to only packages that have at least one package type matching the package type name.
```

To emphasize that _at least one_ must match

File: docs/api/search-query-service-resource.md [hunk 5]
```diff
@@ -119,6 +125,12 @@ Name      | Type    | Required | Notes
 version   | string  | yes      | The full SemVer 2.0.0 version string of the package (could contain build metadata)
 downloads | integer | yes      | The number of downloads for this specific package version
 
+The `packageTypes` array will always consist of at least one (1) item. Package type for a given package id is considered to be the package types defined by the latest version of the package with respect to the other search parameters. Each item in the `packageTypes` array is a JSON object with the following properties:
```
Comment (joelverhagen) [inline docs/api/search-query-service-resource.md:128]:
```suggestion
The `packageTypes` array will always consist of at least one (1) item. Package type for a given package ID considered to be the package types defined by the latest version of the package with respect to the other search parameters. Each item in the `packageTypes` array is a JSON object with the following properties:
```

File: docs/api/search-autocomplete-service-resource.md (hunk 4 above)
Comment (loic-sharma) [inline docs/api/search-autocomplete-service-resource.md:73]: `futher` -> `further`

File: docs/api/catalog-resource.md [hunk 6]
```diff
@@ -227,6 +227,7 @@ minClientVersion        | string                     | no       |
 packageHash             | string                     | yes      | The hash of the package, encoding using [standard base 64](https://tools.ietf.org/html/rfc4648#section-4)
 packageHashAlgorithm    | string                     | yes      |
 packageSize             | integer                    | yes      | The size of the package .nupkg in bytes
+packageTypes            | array of object            | no       | The package types specified by the author.
```
Comment (loic-sharma) [inline docs/api/catalog-resource.md:230]:
```suggestion
packageTypes            | array of objects            | no       | The package types specified by the author.
```

File: docs/api/search-autocomplete-service-resource.md [hunk 7]
```diff
@@ -50,6 +50,7 @@ skip        | URL    | integer | no       | The number of results to skip, for p
 take        | URL    | integer | no       | The number of results to return, for pagination
 prerelease  | URL    | boolean | no       | `true` or `false` determining whether to include [pre-release packages](../create-packages/prerelease-packages.md)
 semVerLevel | URL    | string  | no       | A SemVer 1.0.0 version string 
+packageType | URL    | string  | no       | The package type to use to filter packages
```
Comment (joelverhagen) [inline docs/api/search-autocomplete-service-resource.md:53]: Could you mention `SearchAutocompleteService/3.5.0` and `SearchQueryService/3.5.0` similar to how we mentioned 3.6.0 for for registration?
https://docs.microsoft.com/en-us/nuget/api/registration-base-url-resource#versioning

Basically we add a row to the table and have a dedicated section about new version. We should emphasize that it's a) backwards compatible for `SearchAutocompleteService` and introduces the `packageType` parameter and `packageTypes` response property (just for search, not autocomplete of course)

File: docs/api/search-query-service-resource.md [hunk 8]
```diff
@@ -105,6 +110,7 @@ tags           | string or array of strings | no       |
 title          | string                     | no       | 
 totalDownloads | integer                    | no       | This value can be inferred by the sum of downloads in the `versions` array
 verified       | boolean                    | no       | A JSON boolean indicating whether the package is [verified](../nuget-org/id-prefix-reservation.md)
+packageTypes   | array of objects           | yes      | The package types defined by the package author
```
Comment (loic-sharma) [inline docs/api/search-query-service-resource.md:113]: Most NuGet V3 search implementations (like Azure DevOps, BaGet, Sleet) do not provide the `packageTypes` property. I would either make this not required, or, add in the notes that the property was added by search v3.5.0

File: docs/api/search-query-service-resource.md [hunk 9]
```diff
@@ -52,6 +52,7 @@ skip        | URL    | integer | no       | The number of results to skip, for p
 take        | URL    | integer | no       | The number of results to return, for pagination
 prerelease  | URL    | boolean | no       | `true` or `false` determining whether to include [pre-release packages](../create-packages/prerelease-packages.md)
 semVerLevel | URL    | string  | no       | A SemVer 1.0.0 version string 
+packageType | URL    | string  | no       | The package type to use to filter packages
```
Comment (loic-sharma) [inline docs/api/search-query-service-resource.md:55]: We should mention this was added in search version 3.5.0. Also, the `Versioning` section should mention that search 3.5.0 exists

File: docs/api/catalog-resource.md [hunk 10]
```diff
@@ -244,6 +245,15 @@ time before the catalog item's commit timestamp.
 The `packageHashAlgorithm` is a string defined by the server implementation representing the hashing algorithm used to
 produce the `packageHash`. nuget.org always used the `packageHashAlgorithm` value of `SHA512`.
 
+The `packageTypes` property will only be present if a package type was specified by the author. If it is present, it will always have at least one (1) entry. Each item in the `packageTypes` array is a JSON object with the following properties:
```
Comment (loic-sharma) [inline docs/api/catalog-resource.md:248]: Should we update the sample catalog leaf response to include the `packageTypes` property?

File: docs/api/search-autocomplete-service-resource.md (hunk 7 above)
Comment (loic-sharma) [inline docs/api/search-autocomplete-service-resource.md:53]: We should mention this was added in autocomplete version 3.5.0. Also, the Versioning section should mention that autocomplete version 3.5.0 exists

**Comment by anangaur** [comment]
Docs Build status updates of commit _[5f4121c](https://github.com/NuGet/docs.microsoft.com-nuget/commits/5f4121cceb8dd0d48531712cebdb1da0a74134b5)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/api/_data/catalog-package-details.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/_data/catalog-package-details.json) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/catalog-resource?branch=pr-en-us-2076) |
[docs/api/_data/search-result.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/_data/search-result.json) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-query-service-resource?branch=pr-en-us-2076) |
[docs/api/catalog-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/catalog-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/catalog-resource?branch=pr-en-us-2076) |
[docs/api/search-autocomplete-service-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/search-autocomplete-service-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-autocomplete-service-resource?branch=pr-en-us-2076) |
[docs/api/search-query-service-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/search-query-service-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-query-service-resource?branch=pr-en-us-2076) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2020%5C6%5C23%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C202006230302496633-2076%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=Mx9B%2BsHJbdy7HxIgW34V7ID2f1EiYFI%2FTKEj2a63lPI%3D&st=2020-06-23T02%3A58%3A19Z&se=2020-07-24T03%3A03%3A19Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/19a7edc0-f91d-0fd5-6e5f-4d131db05c11?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

File: docs/api/_data/catalog-package-details.json [hunk 11]
```diff
@@ -28,6 +28,13 @@
   "packageHash": "2edCwKLcbcgFJpsAwa883BLtOy8bZpWwbQpiIb71E74k5t2f2WzXEGWbPwntRleUEgSrcxJrh9Orm/TAmgO4NQ==",
   "packageHashAlgorithm": "SHA512",
   "packageSize": 118348,
+  "packageTypes": [
+    {
+      "@id": "https://api.nuget.org/v3/catalog0/data/2015.02.01.11.18.40/windowsazure.storage.1.0.0.json#packagetypes/DotnetTool",
+      "@type": "PackageType",
+      "name": "DotnetTool"
+    }
+  ]
```
Comment (joelverhagen) [inline docs/api/_data/catalog-package-details.json:37]:
```suggestion
  ],
```

**Comment by anangaur** [comment]
Docs Build status updates of commit _[7eab506](https://github.com/NuGet/docs.microsoft.com-nuget/commits/7eab506f5b8cd9ae0ada9ff07aeeb102d2114644)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/api/_data/catalog-package-details.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/_data/catalog-package-details.json) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/catalog-resource?branch=pr-en-us-2076) |
[docs/api/_data/search-result.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/_data/search-result.json) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-query-service-resource?branch=pr-en-us-2076) |
[docs/api/catalog-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/catalog-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/catalog-resource?branch=pr-en-us-2076) |
[docs/api/search-autocomplete-service-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/search-autocomplete-service-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-autocomplete-service-resource?branch=pr-en-us-2076) |
[docs/api/search-query-service-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/search-query-service-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-query-service-resource?branch=pr-en-us-2076) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2020%5C6%5C23%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C202006230306167961-2076%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=jjHIZKiMgAXLQDDb%2FAqkcG4RjLddaJgz4Wr4NXkfbaI%3D&st=2020-06-23T03%3A01%3A54Z&se=2020-07-24T03%3A06%3A54Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/19a7edc0-f91d-0fd5-6e5f-4d131db05c11?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

**Comment by anangaur** [comment]
Docs Build status updates of commit _[e050cd3](https://github.com/NuGet/docs.microsoft.com-nuget/commits/e050cd3042923094a4660c934e92358e18c87904)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/api/_data/catalog-package-details.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/_data/catalog-package-details.json) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/catalog-resource?branch=pr-en-us-2076) |
[docs/api/_data/search-result.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/_data/search-result.json) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-query-service-resource?branch=pr-en-us-2076) |
[docs/api/catalog-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/catalog-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/catalog-resource?branch=pr-en-us-2076) |
[docs/api/search-autocomplete-service-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/search-autocomplete-service-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-autocomplete-service-resource?branch=pr-en-us-2076) |
[docs/api/search-query-service-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/ryuyu-packageTypes-update/docs/api/search-query-service-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/search-query-service-resource?branch=pr-en-us-2076) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2020%5C6%5C23%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C202006230307139218-2076%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=VCDbXh7M8IUpuxt7T3TX03WY6Lcj0Qy0OYTyrzfpgeA%3D&st=2020-06-23T03%3A02%3A43Z&se=2020-07-24T03%3A07%3A43Z&sp=r).

**Note:** Broken links written as relative paths are included in the above build report. For broken links written as absolute paths or external URLs, see the [broken link report](https://ops.microsoft.com/#/repos/19a7edc0-f91d-0fd5-6e5f-4d131db05c11?tabName=brokenlinks).

For any questions, please:<ul><li>Try searching in the <a href="https://review.docs.microsoft.com/en-us/search/?search=&amp;category=All&amp;scope=help-docs&amp;category=All&amp;branch=master">Docs contributor and Admin Guide</a></li><li>See the <a href="https://review.docs.microsoft.com/en-us/help/onboard/faq?branch=master">frequently asked questions</a></li><li>Post your question in the <a href="https://teams.microsoft.com/l/channel/19%3a7ecffca1166a4a3986fed528cf0870ee%40thread.skype/General?groupId=de9ddba4-2574-4830-87ed-41668c07a1ca&amp;tenantId=72f98bf-86f1-41af-91ab-2d7cd011db47">Docs support channel</a></li></ul>

**Pull request merged**